src/libAtomVM/otp_net.c

```diff
@@ -0,0 +1,310 @@
+        case EAI_BADFLAGS:
+            return globalcontext_make_atom(glb, ATOM_STR("\xB", "eaibadflags"));
+        case EAI_FAIL:
+            return globalcontext_make_atom(glb, ATOM_STR("\x7", "eaifail"));
+        case EAI_FAMILY:
+            return globalcontext_make_atom(glb, ATOM_STR("\x9", "eaifamily"));
+        case EAI_MEMORY:
+            return globalcontext_make_atom(glb, ATOM_STR("\x9", "eaimemory"));
+        case EAI_NONAME:
+            return globalcontext_make_atom(glb, ATOM_STR("\x9", "eainoname"));
+        case EAI_SERVICE:
+            return globalcontext_make_atom(glb, ATOM_STR("\xA", "eaiservice"));
+        case EAI_SOCKTYPE:
+            return globalcontext_make_atom(glb, ATOM_STR("\xB", "eaisocktype"));
+#ifdef HAVE_EXTENDED_EAI_ERRNO
+        case EAI_BADHINTS:
+            return globalcontext_make_atom(glb, ATOM_STR("\xB", "eaibadhints"));
+        case EAI_OVERFLOW:
+            return globalcontext_make_atom(glb, ATOM_STR("\xB", "eaioverflow"));
+        case EAI_PROTOCOL:
+            return globalcontext_make_atom(glb, ATOM_STR("\xB", "eaiprotocol"));
+        case EAI_SYSTEM:
+            return globalcontext_make_atom(glb, ATOM_STR("\x9", "eaisystem"));
+#endif
+    }
+    return term_from_int(err);
+}
+
+static inline term make_address(Context *ctx, struct sockaddr_in *addr)
+{
+    return socket_tuple_from_addr(ctx, ntohl(addr->sin_addr.s_addr));
+}
+
+//
+// net:getaddrinfo/1
+//
+
+static term nif_net_getaddrinfo(Context *ctx, int argc, term argv[])
+{
+    TRACE("nif_net_getaddrinfo\n");
+    UNUSED(argc);
+
+    GlobalContext *global = ctx->global;
+
+    term host = argv[0];
+    term service = argv[1];
+
+    if (host == UNDEFINED_ATOM && service == UNDEFINED_ATOM) {
+        TRACE("Host and Service params may not both be undefined");
+        RAISE_ERROR(BADARG_ATOM);
+    }
+
+    char *host_str = NULL;
+    if (host != UNDEFINED_ATOM) {
+        int ok;
+        host_str = interop_term_to_string(host, &ok);
+        if (!ok) {
+            RAISE_ERROR(BADARG_ATOM);
+        }
+        TRACE("Host: %s\n", host_str);
+    }
+
+    char *service_str = NULL;
+    if (service != UNDEFINED_ATOM) {
+        int ok;
+        service_str = interop_term_to_string(service, &ok);
+        if (!ok) {
+            free(host_str);
+            RAISE_ERROR(BADARG_ATOM);
+        }
+        TRACE("Service: %s\n", service_str);
+    }
+
+    avm_uint_t port = 0;
+#ifdef HAVE_SERVBYNAME
+    if (!IS_NULL_PTR(service_str)) {
+        struct servent *svt = getservbyname(service_str, NULL);
+        if (!IS_NULL_PTR(svt)) {
+            port = ntohs(svt->s_port);
+        }
+    }
+#endif
+    TRACE("port: %zu\n", port);
+
+    // for now, we are only supporting IPv4 addresses
+    struct addrinfo hints;
+    memset(&hints, 0, sizeof(hints));
+    hints.ai_family = PF_INET;
+
+    struct addrinfo *host_info;
+    int err = getaddrinfo(host_str, service_str, &hints, &host_info);
```

Comment:
This hides quite a huge problem: `getaddrinfo` is a blocking function involving network I/O. This might block the caller for quite some time, and this is not an ideal approach inside a NIF.

Reply:
True, but there is no POSIX async version of this function, and I don't think we want to write out own DNS resolver using sockets, for example.

If we were, for example, to implement this call in a port, it would still block the core executing the function, and this would not relenquish control to the scheduler.  Code running on another core could run, but chances are it would just be waiting for the response from this function, anyway, so I don't see any real advantage to that approach.  And in fact it would be pretty complicated, because you might want to spin the `getaddrinfo(3)` call off into a thread or RTOS task, which would involve mutexes or queues or other mechanisms for inter-thread/task communication.

BTW, `erlang.org` _does_ take some time to resolve!

Reply:
Let's merge this first iteration and think later about an async version.